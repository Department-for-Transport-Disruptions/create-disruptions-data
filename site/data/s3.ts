import { S3Client, PutObjectCommand, GetObjectCommand } from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";
import logger from "../utils/logger";

const s3 = new S3Client({ region: "eu-west-2" });

export const putItem = async (
    bucket: string,
    key: string,
    body: string | Blob | Uint8Array | Buffer,
): Promise<void> => {
    logger.info("", {
        context: "data.s3",
        message: "uploading item to s3",
    });

    try {
        const input = {
            Body: body,
            Bucket: bucket,
            Key: key,
        };
        const command = new PutObjectCommand(input);
        await s3.send(command);
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to put item into s3: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const getItem = async (bucket: string, key: string, originalFilename: string): Promise<string> => {
    logger.info("", {
        context: "data.s3",
        message: "getting item from s3",
    });

    try {
        const input = {
            Bucket: bucket,
            Key: key,
            ResponseContentDisposition: 'attachment; filename ="' + originalFilename + '"',
        };
        const command = new GetObjectCommand(input);
        const url = await getSignedUrl(s3, command, { expiresIn: 3600 });
        return url;
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to get item from s3: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const getObject = async (bucket: string, key: string, originalFilename: string): Promise<Uint8Array | null> => {
    logger.info("", {
        context: "data.s3",
        message: "getting item from s3",
    });

    try {
        const input = {
            Bucket: bucket,
            Key: key,
            ResponseContentDisposition: 'attachment; filename ="' + originalFilename + '"',
        };
        const command = new GetObjectCommand(input);
        const response = await s3.send(command);
        return response.Body?.transformToByteArray() ?? null;
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to get item from s3: ${error.stack || ""}`);
        }

        throw error;
    }
};
