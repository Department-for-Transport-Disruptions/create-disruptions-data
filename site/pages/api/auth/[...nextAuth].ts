import NextAuth from "next-auth";
import CognitoProvider from "next-auth/providers/cognito";

export default NextAuth({
    providers: [
        CognitoProvider({
            clientId: "",
            clientSecret: "",
            issuer: "",
        }),
    ],
    pages: {
        signIn: "",
        newUser: "",
    },
    session: {
        strategy: "jwt",
    },
});
