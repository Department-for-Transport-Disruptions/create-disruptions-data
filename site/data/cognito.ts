import {
    CognitoIdentityProviderClient,
    AdminInitiateAuthCommand,
    AdminInitiateAuthCommandInput,
    AdminInitiateAuthCommandOutput,
    AdminRespondToAuthChallengeCommand,
    AdminRespondToAuthChallengeCommandInput,
    AdminUserGlobalSignOutCommand,
    AdminUserGlobalSignOutCommandInput,
    AdminSetUserPasswordCommand,
    AdminSetUserPasswordCommandInput,
    AdminAddUserToGroupCommand,
    ListGroupsCommand,
    ListUsersInGroupCommand,
    UserType,
    AdminDeleteUserCommand,
    AdminGetUserCommand,
    AdminGetUserCommandInput,
    AdminDeleteUserCommandInput,
    AdminCreateUserCommand,
    ListUsersCommand,
    ConfirmForgotPasswordCommand,
    ConfirmForgotPasswordCommandInput,
    ForgotPasswordCommandInput,
    ForgotPasswordCommand,
} from "@aws-sdk/client-cognito-identity-provider";

import { createHmac } from "crypto";
import { AddUserSchema } from "../schemas/add-user.schema";
import { userManagementSchema } from "../schemas/user-management.schema";
import { notEmpty } from "../utils";
import logger from "../utils/logger";

const {
    COGNITO_CLIENT_ID: cognitoClientId,
    COGNITO_CLIENT_SECRET: cognitoClientSecret,
    COGNITO_USER_POOL_ID: userPoolId,
} = process.env;

if (!cognitoClientSecret || !cognitoClientId || !userPoolId) {
    throw new Error("Cognito env vars not set");
}

const cognito = new CognitoIdentityProviderClient({
    region: "eu-west-2",
});

export const deleteUser = (username: string) => {
    try {
        const params: AdminDeleteUserCommandInput = {
            UserPoolId: userPoolId,
            Username: username,
        };
        return cognito.send(new AdminDeleteUserCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to delete user: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const getUserDetails = (username: string) => {
    try {
        const params: AdminGetUserCommandInput = {
            UserPoolId: userPoolId,
            Username: username,
        };
        return cognito.send(new AdminGetUserCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to get user details by username: ${error.stack || ""}`);
        }

        throw error;
    }
};

const calculateSecretHash = (username: string): string =>
    createHmac("SHA256", cognitoClientSecret)
        .update(username + cognitoClientId)
        .digest("base64");

export const initiateAuth = async (username: string, password: string): Promise<AdminInitiateAuthCommandOutput> => {
    logger.info("", {
        context: "data.cognito",
        message: "initiating auth",
    });

    try {
        const params: AdminInitiateAuthCommandInput = {
            AuthFlow: "ADMIN_USER_PASSWORD_AUTH",
            ClientId: cognitoClientId,
            UserPoolId: userPoolId,
            AuthParameters: {
                USERNAME: username,
                PASSWORD: password,
                SECRET_HASH: calculateSecretHash(username),
            },
        };

        return cognito.send(new AdminInitiateAuthCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to authenticate user: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const respondToNewPasswordChallenge = async (
    username: string,
    password: string,
    session: string,
): Promise<void> => {
    logger.info("", {
        context: "data.cognito",
        message: "new password challenge initiated",
    });

    try {
        const params: AdminRespondToAuthChallengeCommandInput = {
            ChallengeName: "NEW_PASSWORD_REQUIRED",
            ClientId: cognitoClientId,
            UserPoolId: userPoolId,
            ChallengeResponses: {
                USERNAME: username,
                NEW_PASSWORD: password,
                SECRET_HASH: calculateSecretHash(username),
            },
            Session: session,
        };

        await cognito.send(new AdminRespondToAuthChallengeCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to respond to password challenge: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const globalSignOut = async (username: string): Promise<void> => {
    logger.info("", {
        context: "data.cognito",
        message: "performing global sign out",
    });

    const params: AdminUserGlobalSignOutCommandInput = {
        Username: username,
        UserPoolId: userPoolId,
    };

    try {
        await cognito.send(new AdminUserGlobalSignOutCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to perform global sign out: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const updateUserPassword = async (newPassword: string, username: string): Promise<void> => {
    logger.info("", {
        context: "data.cognito",
        message: "updating user password",
    });

    const params: AdminSetUserPasswordCommandInput = {
        Password: newPassword,
        Permanent: true,
        Username: username,
        UserPoolId: userPoolId,
    };

    try {
        await cognito.send(new AdminSetUserPasswordCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to update user password: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const getAllUsersInGroup = async (groupName: string, nextToken?: string): Promise<UserType[]> => {
    const users = await cognito.send(
        new ListUsersInGroupCommand({ UserPoolId: userPoolId, GroupName: groupName, Limit: 60, NextToken: nextToken }),
    );

    if (!users.Users) {
        return [];
    }

    if (users.NextToken) {
        return [...users.Users, ...(await getAllUsersInGroup(groupName, users.NextToken))];
    } else {
        return users.Users;
    }
};

export const listUsersWithGroups = async () => {
    logger.info("", {
        context: "data.cognito",
        message: "Listing cognito users",
    });

    try {
        const groupList = await cognito.send(new ListGroupsCommand({ UserPoolId: userPoolId }));

        const userList = await Promise.all(
            groupList.Groups?.map(async (group) => {
                if (!group.GroupName) {
                    return null;
                }

                const groupUsers = await getAllUsersInGroup(group.GroupName);

                return groupUsers.map((user) => ({ ...user, GroupName: group.GroupName }));
            }) ?? [],
        );

        const parsedUsers = userManagementSchema.safeParse(userList.filter(notEmpty).flat());

        if (!parsedUsers.success) {
            return [];
        }

        return parsedUsers.data;
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to list cognito users: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const deleteUsersByAttribute = async (attributeName: string, attributeValue: string) => {
    logger.info("", {
        context: "data.cognito",
        message: "Listing cognito users based on custom attributes and deleting the users that match the value",
    });

    try {
        const userList = await cognito.send(
            new ListUsersCommand({ UserPoolId: userPoolId, AttributesToGet: [attributeName] }),
        );

        const deletePromises = userList.Users?.filter((user) =>
            user.Attributes?.find((value) => value.Value === attributeValue),
        ).map((user) => {
            return cognito.send(
                new AdminDeleteUserCommand({
                    UserPoolId: userPoolId,
                    Username: user.Username,
                }),
            );
        });

        await Promise.all(deletePromises ?? []);
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(
                `Failed to list cognito users based on attribute ${attributeName} with value ${attributeValue}: ${
                    error.stack || ""
                }`,
            );
        }

        throw error;
    }
};

export const createUser = async (userData: AddUserSchema) => {
    logger.info("", {
        context: "data.cognito",
        message: "Adding a new user",
    });

    const nocCodes = userData.operatorNocInfo?.map((operator) => operator.nocCode) ?? [];

    const userAttributes = [
        {
            Name: "custom:orgId",
            Value: userData.orgId,
        },
        {
            Name: "given_name",
            Value: userData.givenName,
        },
        {
            Name: "family_name",
            Value: userData.familyName,
        },
        {
            Name: "email_verified",
            Value: "true",
        },
        {
            Name: "email",
            Value: userData.email,
        },
    ];

    const createUserCommand = {
        Username: userData.email,
        UserPoolId: userPoolId,
        TemporaryPassword: Array.from(Array(20), () => Math.floor(Math.random() * 36).toString(36)).join(""),
        UserAttributes: userAttributes,
    };

    const createOperatorUserCommand = {
        ...createUserCommand,
        UserAttributes: userAttributes.concat([
            {
                Name: "custom:nocCodes",
                Value: nocCodes.toString(),
            },
        ]),
    };

    const createUserResult =
        userData.group === "operators"
            ? await cognito.send(new AdminCreateUserCommand(createOperatorUserCommand))
            : await cognito.send(new AdminCreateUserCommand(createUserCommand));

    await cognito.send(
        new AdminAddUserToGroupCommand({
            GroupName: userData.group,
            Username: createUserResult.User?.Username,
            UserPoolId: userPoolId,
        }),
    );
};

export const initiateResetPassword = async (email: string) => {
    logger.info("", {
        context: "data.cognito",
        message: "Initiating reset password flow for cognito user",
    });

    try {
        const params: ForgotPasswordCommandInput = {
            Username: email,
            ClientId: cognitoClientId,
            SecretHash: calculateSecretHash(email),
        };
        return cognito.send(new ForgotPasswordCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to initiate reset password flow: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const resetUserPassword = async (key: string, newPassword: string, email: string) => {
    logger.info("", {
        context: "data.cognito",
        message: "Resetting password for cognito user",
    });

    try {
        const params: ConfirmForgotPasswordCommandInput = {
            ClientId: cognitoClientId,
            ConfirmationCode: key,
            Password: newPassword,
            Username: email,
            SecretHash: calculateSecretHash(email),
        };
        return cognito.send(new ConfirmForgotPasswordCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to reset password: ${error.stack || ""}`);
        }

        throw error;
    }
};
