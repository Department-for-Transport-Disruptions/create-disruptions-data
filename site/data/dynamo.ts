import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocumentClient, QueryCommand } from "@aws-sdk/lib-dynamodb";
import { ptSituationElementSchema } from "@create-disruptions-data/shared-ts/siriTypes.zod";
import { notEmpty } from "../utils";
import logger from "../utils/logger";

const ddbDocClient = DynamoDBDocumentClient.from(new DynamoDBClient({ region: "eu-west-2" }));

const tableName = process.env.TABLE_NAME as string;

export const getDisruptionsDataFromDynamo = async () => {
    logger.info("Getting disruptions data from DynamoDB table...");

    const dbData = await ddbDocClient.send(
        new QueryCommand({
            TableName: tableName,
            KeyConditionExpression: "PK = :i",
            ExpressionAttributeValues: {
                ":i": "1",
            },
        }),
    );

    return dbData.Items?.map((item) => {
        const parsedItem = ptSituationElementSchema.safeParse(item);

        if (!parsedItem.success) {
            logger.error("Error parsing disruption from dynamo");
            logger.error(parsedItem.error.stack);
            return null;
        }

        return parsedItem.data;
    }).filter(notEmpty);
};
