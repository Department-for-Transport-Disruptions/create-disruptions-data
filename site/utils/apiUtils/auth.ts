import { decodeJwt } from "jose";
import { NextApiRequest } from "next";
import { parseCookies } from "nookies";
import { z } from "zod";
import { IncomingMessage } from "http";
import { COOKIES_ID_TOKEN } from "../../constants";
import { getOrganisationInfoById } from "../../data/dynamo";

const sessionSchema = z
    .object({
        sub: z.string().uuid(),
        email: z.string().email(),
        "custom:orgId": z.string().uuid(),
        "cognito:groups": z.array(
            z.union([
                z.literal("system-admins"),
                z.literal("org-admins"),
                z.literal("org-publishers"),
                z.literal("org-staff"),
            ]),
        ),
    })
    .transform((item) => {
        const isSystemAdmin = item["cognito:groups"].includes("system-admins");
        const isOrgAdmin = isSystemAdmin || item["cognito:groups"].includes("org-admins");
        const isOrgPublisher = isSystemAdmin || isOrgAdmin || item["cognito:groups"].includes("org-publishers");
        const isOrgStaff =
            isSystemAdmin || isOrgAdmin || isOrgPublisher || item["cognito:groups"].includes("org-staff");

        return {
            username: item.sub,
            email: item.email,
            orgId: item["custom:orgId"],
            isSystemAdmin,
            isOrgAdmin,
            isOrgPublisher,
            isOrgStaff,
        };
    });

const sessionSchemaWithOrgDetail = sessionSchema.transform(async (item) => {
    const orgDetail = await getOrganisationInfoById(item.orgId);

    return {
        ...item,
        orgName: orgDetail?.name ?? "",
        adminAreaCodes: orgDetail?.adminAreaCodes ?? [],
    };
});

export type Session = z.infer<typeof sessionSchema>;
export type SessionWithOrgDetail = z.infer<typeof sessionSchemaWithOrgDetail>;

export const getSession = (req: NextApiRequest | IncomingMessage): Session | null => {
    const cookies = parseCookies({ req });
    const idToken = cookies[COOKIES_ID_TOKEN];

    if (idToken) {
        return sessionSchema.parse(decodeJwt(idToken));
    }

    return null;
};

export const getSessionWithOrgDetail = async (
    req: NextApiRequest | IncomingMessage,
): Promise<SessionWithOrgDetail | null> => {
    const cookies = parseCookies({ req });
    const idToken = cookies[COOKIES_ID_TOKEN];

    if (idToken) {
        return sessionSchemaWithOrgDetail.parseAsync(decodeJwt(idToken));
    }

    return null;
};
