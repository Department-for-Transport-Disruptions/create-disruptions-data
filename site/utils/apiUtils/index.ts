import { Consequence } from "@create-disruptions-data/shared-ts/disruptionTypes";
import { MAX_CONSEQUENCES } from "@create-disruptions-data/shared-ts/disruptionTypes.zod";
import { SocialMediaPostStatus } from "@create-disruptions-data/shared-ts/enums";
import { NextApiRequest, NextApiResponse } from "next";
import { parseCookies, setCookie } from "nookies";
import { TwitterApi } from "twitter-api-v2";
import { z } from "zod";
import { IncomingMessage, ServerResponse } from "http";
import { notEmpty } from "..";
import {
    COOKIES_POLICY_COOKIE,
    COOKIE_CSRF,
    COOKIES_ID_TOKEN,
    COOKIE_PREFERENCES_COOKIE,
    DISRUPTION_DETAIL_PAGE_PATH,
    REVIEW_DISRUPTION_PAGE_PATH,
    COOKIES_REFRESH_TOKEN,
} from "../../constants";
import { upsertConsequence } from "../../data/dynamo";
import { getAccessToken, publishToHootsuite } from "../../data/hootsuite";
import { getAuthedTwitterClient, sendTweet } from "../../data/twitter";
import { TooManyConsequencesError } from "../../errors";
import { PageState } from "../../interfaces";
import { SocialMediaPost } from "../../schemas/social-media.schema";
import logger from "../logger";

export const setCookieOnResponseObject = (
    cookieName: string,
    cookieValue: string,
    res: NextApiResponse | ServerResponse<IncomingMessage>,
    lifetime?: number,
    httpOnly = true,
    sameSite: "lax" | "strict" | "none" = "lax",
): void => {
    // From docs: All cookies are httponly by default, and cookies sent over SSL are secure by
    // default. An error will be thrown if you try to send secure cookies over an insecure socket.
    setCookie({ res }, cookieName, cookieValue, {
        path: "/",
        sameSite,
        secure: process.env.NODE_ENV !== "development",
        maxAge: lifetime,
        httpOnly,
    });
};

export const destroyCookieOnResponseObject = (
    cookieName: string,
    res: NextApiResponse | ServerResponse<IncomingMessage>,
): void => {
    setCookieOnResponseObject(cookieName, "", res, 0);
};

export const redirectTo = (res: NextApiResponse | ServerResponse, location: string): void => {
    res.writeHead(302, {
        Location: location,
    });
    res.end();
};

export const redirectToError = (
    res: NextApiResponse | ServerResponse,
    message?: string,
    context?: string,
    error?: Error,
): void => {
    if (message && context && error) {
        logger.error(message, { context, error: error.stack });
    }

    redirectTo(res, "/500");
};

export const cleardownCookies = (req: NextApiRequest, res: NextApiResponse) => {
    const cookies = parseCookies({ req });

    const saveList = [
        COOKIES_POLICY_COOKIE,
        COOKIE_PREFERENCES_COOKIE,
        COOKIES_ID_TOKEN,
        COOKIES_REFRESH_TOKEN,
        COOKIE_CSRF,
    ];

    Object.keys(cookies).forEach((cookie) => {
        if (!saveList.includes(cookie)) {
            destroyCookieOnResponseObject(cookie, res);
        }
    });
};

export const getPageState = <T>(errorCookie: string, schemaObject: z.ZodType<T>, disruptionId?: string, data?: T) => {
    const inputsProps: PageState<Partial<T>> = {
        errors: [],
        inputs: {},
        disruptionId: disruptionId || "",
    };

    if (errorCookie) {
        return { ...(JSON.parse(errorCookie) as PageState<Partial<T>>), disruptionId: disruptionId || "" };
    }

    if (disruptionId) {
        const parsedData = schemaObject.safeParse(data);

        if (parsedData.success) {
            inputsProps.inputs = parsedData.data;
        }
    }

    return inputsProps;
};

export const getReturnPage = (req: NextApiRequest) => {
    const queryParam = req.headers.referer?.split("?")[1];
    const decodedQueryParam = queryParam ? decodeURIComponent(queryParam) : null;
    return decodedQueryParam?.includes(REVIEW_DISRUPTION_PAGE_PATH) ||
        decodedQueryParam?.includes(DISRUPTION_DETAIL_PAGE_PATH)
        ? queryParam
        : null;
};

export const isDisruptionFromTemplate = (req: NextApiRequest) => {
    const queryParam = req.headers.referer?.split("?")[1];

    const decodedQueryParam = queryParam ? decodeURIComponent(queryParam) : null;
    return decodedQueryParam?.includes(DISRUPTION_DETAIL_PAGE_PATH) && decodedQueryParam.includes("template=true")
        ? queryParam
        : null;
};

export const delay = (ms: number) => new Promise((res) => setTimeout(res, ms));

export const publishSocialMedia = async (
    socialMediaPosts: SocialMediaPost[],
    orgId: string,
    isUserStaff: boolean,
    canPublish: boolean,
) => {
    const authedTwitterClients: Record<string, TwitterApi> = {};
    const hootsuiteAccessTokens: Record<string, string> = {};

    const uniqueTwitterSocialAccounts = new Set(
        socialMediaPosts.filter((post) => post.accountType === "Twitter").map((post) => post.socialAccount),
    );

    const uniqueHootsuiteSocialAccounts = new Set(
        socialMediaPosts.filter((post) => post.accountType === "Hootsuite").map((post) => post.socialAccount),
    );

    for (const socialAccount of uniqueHootsuiteSocialAccounts) {
        const accessToken = await getAccessToken(orgId, socialAccount);
        hootsuiteAccessTokens[socialAccount] = accessToken;
    }

    for (const socialAccount of uniqueTwitterSocialAccounts) {
        const authedClient = await getAuthedTwitterClient(orgId, socialAccount);

        if (authedClient) {
            authedTwitterClients[socialAccount] = authedClient;
        }
    }

    const socialMediaPromises = socialMediaPosts.map((post) => {
        if (post.status === SocialMediaPostStatus.pending) {
            if (post.accountType === "Twitter") {
                return sendTweet(orgId, post, isUserStaff, canPublish, authedTwitterClients[post.socialAccount]);
            }

            return publishToHootsuite(post, orgId, isUserStaff, canPublish, hootsuiteAccessTokens[post.socialAccount]);
        }

        return null;
    });

    await Promise.all(socialMediaPromises);
};

export const redirectToWithQueryParams = (
    req: NextApiRequest,
    res: NextApiResponse,
    queryParamsToForward: string[],
    location: string,
    paramsToAdd?: string[],
) => {
    const queryStringParams = queryParamsToForward
        .map((p) => {
            const paramValue = req.query[p];

            return paramValue ? `${p}=${paramValue.toString()}` : null;
        })
        .filter(notEmpty);

    redirectTo(
        res,
        `${location}${
            [...queryStringParams, ...(paramsToAdd || [])].length > 0
                ? `?${[...queryStringParams, ...(paramsToAdd || [])].join("&")}`
                : ""
        }`,
    );
};

export const handleUpsertConsequence = async (
    consequence: Consequence | Pick<Consequence, "disruptionId" | "consequenceIndex">,
    orgId: string,
    isOrgStaff: boolean,
    isTemplate: boolean,
    inputs: unknown,
    errorCookie: string,
    res: NextApiResponse,
) => {
    try {
        return await upsertConsequence(consequence, orgId, isOrgStaff, isTemplate);
    } catch (e) {
        if (e instanceof TooManyConsequencesError) {
            setCookieOnResponseObject(
                errorCookie,
                JSON.stringify({
                    inputs,
                    errors: [
                        {
                            id: "",
                            errorMessage: `Max consequence limit of ${MAX_CONSEQUENCES} has been reached`,
                        },
                    ],
                }),
                res,
            );
        }

        throw e;
    }
};
