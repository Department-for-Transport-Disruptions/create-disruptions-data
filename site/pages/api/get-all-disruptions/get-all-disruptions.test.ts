import { PublishStatus } from "@create-disruptions-data/shared-ts/enums";
import MockDate from "mockdate";
import { describe, it, expect, afterEach, beforeEach, vi, afterAll } from "vitest";
import getAllDisruptions, { getDisruptionsForTable } from "./[organisationId].api";
import * as s3 from "../../../data/s3";
import {
    DEFAULT_ORG_ID,
    disruptionArray,
    disruptionWithConsequencesAndSocialMediaPosts,
    getMockRequestAndResponse,
    mockSession,
} from "../../../testData/mockData";
import * as session from "../../../utils/apiUtils/auth";

describe("getAllDisruptions", () => {
    const writeHeadMock = vi.fn();

    vi.mock("../../data/dynamo", () => ({
        getDisruptionsDataFromDynamo: vi.fn(),
    }));

    vi.mock("../../../data/s3");

    const getDisruptionsSpy = vi.spyOn(s3, "getObject");
    const getSessionSpy = vi.spyOn(session, "getSession");

    beforeEach(() => {
        getSessionSpy.mockImplementation(() => {
            return mockSession;
        });
    });

    afterEach(() => {
        vi.resetAllMocks();
    });

    afterAll(() => {
        MockDate.reset();
    });

    it("should be successful when session is set", async () => {
        getDisruptionsSpy.mockResolvedValue("");

        const { req, res } = getMockRequestAndResponse({
            query: {
                organisationId: DEFAULT_ORG_ID,
            },
            mockWriteHeadFn: writeHeadMock,
        });

        res.status = vi.fn().mockImplementation(() => ({
            json: vi.fn(),
        }));

        await getAllDisruptions(req, res);

        expect(getDisruptionsSpy).toHaveBeenCalledOnce();

        expect(res.status).toHaveBeenCalledWith(200);
    });

    it("should return 403 when no session set", async () => {
        getSessionSpy.mockReturnValue(null);

        const { req, res } = getMockRequestAndResponse({
            body: {},
            mockWriteHeadFn: writeHeadMock,
        });

        res.status = vi.fn().mockImplementation(() => ({
            json: vi.fn(),
        }));

        await getAllDisruptions(req, res);

        expect(getDisruptionsSpy).not.toHaveBeenCalledOnce();

        expect(res.status).toHaveBeenCalledWith(403);
    });

    it("should return 403 if orgIds do not match", async () => {
        getDisruptionsSpy.mockResolvedValue("");

        const { req, res } = getMockRequestAndResponse({
            query: {
                organisationId: "123",
            },
            mockWriteHeadFn: writeHeadMock,
        });

        res.status = vi.fn().mockImplementation(() => ({
            json: vi.fn(),
        }));

        await getAllDisruptions(req, res);

        expect(getDisruptionsSpy).not.toHaveBeenCalledOnce();

        expect(res.status).toHaveBeenCalledWith(403);
    });

    describe("getDashboardDisruptions", () => {
        it("should correctly parse disruptions", async () => {
            getDisruptionsSpy.mockResolvedValue(JSON.stringify(disruptionArray));

            const disruptions = await getDisruptionsForTable(DEFAULT_ORG_ID, false);

            expect(getDisruptionsSpy).toHaveBeenCalledOnce();

            expect(disruptions).toMatchSnapshot();
        });

        it("should filter out duplicate disruptions", async () => {
            const disruptionsToStringify = [
                ...disruptionArray,
                {
                    ...disruptionWithConsequencesAndSocialMediaPosts,
                    disruptionId: "ca090776-57c6-46a1-a03a-6e0236ee17c8",
                },
            ];
            disruptionsToStringify[2].publishStatus = PublishStatus.pendingAndEditing;

            getDisruptionsSpy.mockResolvedValue(JSON.stringify(disruptionsToStringify));

            const disruptions = await getDisruptionsForTable(DEFAULT_ORG_ID, false);

            expect(getDisruptionsSpy).toHaveBeenCalledOnce();

            expect(disruptions.length).toBe(3);
        });
    });
});
