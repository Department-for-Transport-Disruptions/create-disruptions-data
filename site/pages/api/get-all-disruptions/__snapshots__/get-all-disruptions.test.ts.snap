// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getAllDisruptions > getDashboardDisruptions > should correctly parse disruptions 1`] = `
[
  {
    "consequenceLength": 2,
    "creationTime": undefined,
    "dataSource": undefined,
    "description": "Test description",
    "displayId": "8fg3ha",
    "disruptionReason": "staffInWrongPlace",
    "disruptionType": "planned",
    "history": undefined,
    "id": "acde070d-8c4c-4f0d-9d8a-162843c10333",
    "isLive": true,
    "isNetworkWideCq": true,
    "isOperatorWideCq": true,
    "modes": [
      "Tram",
      "Bus",
    ],
    "operators": [
      "FSYO",
    ],
    "publishEndDate": "",
    "publishStartDate": "2023-03-10T12:00:00.000Z",
    "services": [],
    "servicesAffectedCount": 0,
    "severity": "severe",
    "status": "draft",
    "stopsAffectedCount": 0,
    "summary": "Some summary",
    "validityPeriods": [
      {
        "endTime": "2023-03-17T17:00:00.000Z",
        "startTime": "2023-03-10T12:00:00.000Z",
      },
      {
        "endTime": null,
        "startTime": "2023-03-18T12:00:00.000Z",
      },
    ],
  },
  {
    "consequenceLength": 0,
    "creationTime": undefined,
    "dataSource": undefined,
    "description": "Test description",
    "displayId": "8fg3ha",
    "disruptionReason": "staffInWrongPlace",
    "disruptionType": "planned",
    "history": undefined,
    "id": "8e9b35a8-1c72-453d-93d5-27983bbf4eba",
    "isLive": true,
    "isNetworkWideCq": false,
    "isOperatorWideCq": false,
    "modes": [],
    "operators": [],
    "publishEndDate": "",
    "publishStartDate": "2023-03-10T12:00:00.000Z",
    "services": [],
    "servicesAffectedCount": 0,
    "severity": "unknown",
    "status": "draft",
    "stopsAffectedCount": 0,
    "summary": "Some summary",
    "validityPeriods": [
      {
        "endTime": "2023-03-17T17:00:00.000Z",
        "startTime": "2023-03-10T12:00:00.000Z",
      },
      {
        "endTime": null,
        "startTime": "2023-03-18T12:00:00.000Z",
      },
    ],
  },
  {
    "consequenceLength": 2,
    "creationTime": undefined,
    "dataSource": undefined,
    "description": "Test description",
    "displayId": "8fg3ha",
    "disruptionReason": "staffInWrongPlace",
    "disruptionType": "planned",
    "history": undefined,
    "id": "ca090776-57c6-46a1-a03a-6e0236ee17c8",
    "isLive": false,
    "isNetworkWideCq": true,
    "isOperatorWideCq": true,
    "modes": [
      "Tram",
      "Bus",
    ],
    "operators": [
      "FSYO",
    ],
    "publishEndDate": "",
    "publishStartDate": "2023-03-10T12:00:00.000Z",
    "services": [],
    "servicesAffectedCount": 0,
    "severity": "severe",
    "status": "open",
    "stopsAffectedCount": 0,
    "summary": "Some summary",
    "validityPeriods": [
      {
        "endTime": null,
        "startTime": "2024-03-10T10:00:00.000Z",
      },
    ],
  },
]
`;
